platform/mellanox/mlnx-platform-api/sonic_platform/modules_mgmt.py

```diff
@@ -0,0 +1,756 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import threading
+import time
+import queue
+import os
+import select
+import traceback
+
+try:
+    from sonic_py_common.logger import Logger
+    from sonic_py_common import device_info, multi_asic
+    from .device_data import DeviceDataManager
+    from sonic_platform_base.sonic_xcvr.fields import consts
+    from sonic_platform_base.sonic_xcvr.api.public import cmis
+    from . import sfp as sfp_module
+    from . import utils
+    from swsscommon.swsscommon import SonicV2Connector
+except ImportError as e:
+    raise ImportError (str(e) + "- required module not found")
+
+# Global logger class instance
+logger = Logger()
+
+STATE_HW_NOT_PRESENT = "Initial state. module is not plugged to cage."
+STATE_HW_PRESENT = "Module is plugged to cage"
+STATE_MODULE_AVAILABLE = "Module hw present and power is good"
+STATE_POWERED = "Module power is already loaded"
+STATE_NOT_POWERED = "Module power is not loaded"
+STATE_FW_CONTROL = "The module is not CMIS and FW needs to handle"
+STATE_SW_CONTROL = "The module is CMIS and SW needs to handle"
+STATE_ERROR_HANDLER = "An error occurred - read/write error, power limit or power cap."
+STATE_POWER_LIMIT_ERROR = "The cage has not enough power for the plugged module"
+STATE_SYSFS_ERROR = "An error occurred while writing/reading SySFS."
+
+SAI_PROFILE_FILE = "/{}/sai.profile"
+SAI_INDEP_MODULE_MODE = "SAI_INDEPENDENT_MODULE_MODE"
+SAI_INDEP_MODULE_MODE_DELIMITER = "="
+SAI_INDEP_MODULE_MODE_TRUE_STR = "1"
+SYSFS_LEGACY_FD_PRESENCE = "/sys/module/sx_core/asic0/module{}/present"
+ASIC_NUM = 0
+SYSFS_INDEPENDENT_FD_PREFIX_WO_MODULE = "/sys/module/sx_core/asic{}".format(ASIC_NUM)
+SYSFS_INDEPENDENT_FD_PREFIX = SYSFS_INDEPENDENT_FD_PREFIX_WO_MODULE + "/module{}"
+SYSFS_INDEPENDENT_FD_PRESENCE = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "hw_present")
+SYSFS_INDEPENDENT_FD_POWER_GOOD = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "power_good")
+SYSFS_INDEPENDENT_FD_POWER_ON = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "power_on")
+SYSFS_INDEPENDENT_FD_HW_RESET = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "hw_reset")
+SYSFS_INDEPENDENT_FD_POWER_LIMIT = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "power_limit")
+SYSFS_INDEPENDENT_FD_FW_CONTROL = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "control")
+# echo <val>  /sys/module/sx_core/$asic/$module/frequency   //  val: 0 - up to 400KHz, 1 - up to 1MHz
+SYSFS_INDEPENDENT_FD_FREQ = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "frequency")
+SYSFS_INDEPENDENT_FD_FREQ_SUPPORT = os.path.join(SYSFS_INDEPENDENT_FD_PREFIX, "frequency_support")
+IS_INDEPENDENT_MODULE = 'is_independent_module'
+
+STATE_DB_TABLE_NAME_PREFIX = 'TRANSCEIVER_MODULES_MGMT|{}'
+
+MAX_EEPROM_ERROR_RESET_RETRIES = 4
+
+class ModulesMgmtTask(threading.Thread):
```

Comment:
@dbarashinvd can you elaborate more on this thread in the PR description, please?

Reply:
this is the main thread of this file, which does the state machine per port.
first the static detection that takes place once the thread is up (during switch bootup sequence).
and after it ends the dynamic detection takes place, listening to changes in the sysfs fds, per port.

Reply:
I added this info also to the PR description.
please note that I updated the PR to fix some issues found recently.